VideoWeb/VideoWeb/ClientApp/src/app/services/user-media-stream.service.spec.ts

```diff
@@ -9,29 +9,44 @@ import { mustProvideAMicrophoneDeviceError, UserMediaStreamService } from './use
 import { UserMediaService } from './user-media.service';
 
 describe('UserMediaStreamService', () => {
+    let cloneStream: any;
     const mediaStreamBuilder = (device: UserMediaDevice) => {
-        const stream = jasmine.createSpyObj<MediaStream>(['addTrack', 'removeTrack', 'getTracks']);
+        const stream = jasmine.createSpyObj<MediaStream>(['addTrack', 'removeTrack', 'getTracks', 'clone']);
         const track = jasmine.createSpyObj<MediaStreamTrack>(['stop'], ['label', 'id']);
-
         getSpiedPropertyGetter(track, 'label').and.returnValue(device.label);
         getSpiedPropertyGetter(track, 'id').and.returnValue(Guid.create().toString());
 
-        stream.getTracks.and.returnValue([track]);
+        const tracks: any = [track];
+        stream.addTrack.and.callFake(x => tracks.push(x));
+
+        stream.removeTrack.and.callFake(x => {
+            const idx = tracks.findIndex(existingTrack => existingTrack.id === x.id);
+            if (idx < 0) {
+                throw new Error(`Cant find track ${track}`);
+            }
+            tracks.splice(idx, 1);
+        });
+
+        stream.getTracks.and.returnValue(tracks);
 
+        stream.clone.and.callFake(() => {
+            cloneStream = Object.assign({}, stream);
+            return cloneStream;
+        });
         return stream;
     };
 
     const cameraOneDevice = new UserMediaDevice('Camera 1', Guid.create().toString(), 'videoinput', '');
-    const cameraOneStream = mediaStreamBuilder(cameraOneDevice);
+    let cameraOneStream = mediaStreamBuilder(cameraOneDevice);
```

Comment:
You could just declare these here since they are assigned to in the beforeEach.

Reply:
moved it to the beforeEach